op-supervisor/supervisor/backend/cross/hazard_set_test.go

```diff
@@ -0,0 +1,802 @@
+				makeBlock(1, 100, 1),
+				makeBlock(1, 100, 2),
+			},
+			expectErr: fmt.Errorf("already depend on"),
+		},
+		{
+			name: "Hazards Across Multiple Chains",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1), makeMessage(2, 100, 1, 1)),
+				makeBlock(1, 100, 1),
+				makeBlock(2, 100, 1),
+			},
+			expected: map[types.ChainIndex]types.BlockSeal{
+				1: makeBlockSeal(1, 100, 1),
+				2: makeBlockSeal(1, 100, 2),
+			},
+		},
+		{
+			name: "Recursive Hazards",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1)),
+				makeBlock(1, 100, 1, makeMessage(2, 100, 1, 1)),
+				makeBlock(2, 100, 1),
+			},
+			expected: map[types.ChainIndex]types.BlockSeal{
+				1: makeBlockSeal(1, 100, 1),
+				2: makeBlockSeal(1, 100, 2),
+			},
+		},
+		{
+			name: "Recursive Hazards - Missing Intermediate Block",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1)),
+				// Block 1 in Chain 1 is missing
+				makeBlock(2, 100, 1),
+			},
+			expectErr: types.ErrFuture,
+		},
+		{
+			name: "Invalid Timestamp - Future Message",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 200, 1, 1)), // Message timestamp > block timestamp
+				makeBlock(1, 100, 1),
+			},
+			expectErr: fmt.Errorf("breaks timestamp invariant"),
+		},
+		{
+			name: "Invalid Timestamp - Zero",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 0, 1, 1)),
+				makeBlock(1, 100, 1),
+			},
+			expectErr: types.ErrFuture,
+		},
+		{
+			name: "Missing Block - Message References Non-existent Block",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 999, 1)), // Block 999 doesn't exist
+			},
+			expectErr: types.ErrFuture,
+		},
+		{
+			name: "Missing Block - Chain Break",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1)),
+				makeBlock(1, 100, 1, makeMessage(2, 100, 1, 1)), // Message references block in chain 2 that doesn't exist
+			},
+			expectErr: types.ErrFuture,
+		},
+		{
+			name: "Invalid Block Number - Zero",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 0, 1)), // Invalid block number
+			},
+			expectErr: types.ErrFuture,
+		},
+		{
+			name: "Recursive Hazards - Diamond Pattern",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1), makeMessage(2, 100, 1, 1)),
+				makeBlock(1, 100, 1, makeMessage(3, 100, 1, 1)),
+				makeBlock(2, 100, 1, makeMessage(3, 100, 1, 1)),
+				makeBlock(3, 100, 1),
+			},
+			expected: map[types.ChainIndex]types.BlockSeal{
+				1: makeBlockSeal(1, 100, 1),
+				2: makeBlockSeal(1, 100, 2),
+				3: makeBlockSeal(1, 100, 3),
+			},
+		},
+		{
+			name: "Multiple Independent Chains",
+			blocks: []blockDef{
+				makeBlock(0, 100, 1, makeMessage(1, 100, 1, 1), makeMessage(2, 100, 1, 1)),
+				// Chain 1: A -> B -> C
+				makeBlock(1, 100, 1, makeMessage(3, 100, 1, 1)),
+				makeBlock(3, 100, 1),
+				// Chain 2: X -> Y -> Z
```

Comment:
are these comments accurate? looks to me like what is actually happening is

// Chain 0 -> Chain 2 -> Chain 4
...
// Chain 1 -> Chain 3